modules/infra/datapath/control_input.c

```diff
@@ -70,13 +77,22 @@ static uint16_t control_input_process(
 	mp = node->ctx_ptr;
 
 	for (unsigned i = 0; i < n; i++) {
-		mbuf = rte_pktmbuf_alloc(mp);
-		if (mbuf) {
-			if (gr_trace_all_enabled())
-				gr_mbuf_trace_add(mbuf, node, 0); // no data
-			control_input_mbuf_data(mbuf)->data = msg[i].data;
-			rte_node_enqueue_x1(graph, node, control_input_edges[msg[i].type], mbuf);
+		if (control_input_edges[msg[i].type].data_is_mbuf) {
+			mbuf = msg[i].data;
+		} else {
+			mbuf = rte_pktmbuf_alloc(mp);
+			if (mbuf)
+				control_input_mbuf_data(mbuf)->data = msg[i].data;
+			else {
+				LOG(ERR, "rte_pktmbuf_alloc %s", rte_strerror(rte_errno));
+				continue;
```

Comment:
More readable to invert this test.

if (!mbuf) {
    LOG(ERR, "rte_pktmbuf_alloc %s", rte_strerror(rte_errno));
    continue;
}

control_input_mbuf_data(mbuf)->data = msg[i].data;

Reply:
done